pkg/frontend/querymiddleware/model_extra.go

```diff
@@ -419,3 +422,47 @@ func (resp *PrometheusResponse) minTime() int64 {
 	}
 	return result[0].Samples[0].TimestampMs
 }
+
+func EncodeCachedHTTPResponse(cacheKey string, res *http.Response) (*CachedHTTPResponse, error) {
+	// Read the response.
+	body, err := readResponseBody(res)
+	if err != nil {
+		return nil, err
+	}
+
+	// Since we've already consumed the response Body we have to replace it on the response,
+	// otherwise the caller will get a response with a closed Body.
+	res.Body = io.NopCloser(bytes.NewBuffer(body))
+
+	// When preallocating the slice we assume that header as 1 value (which is the most common case).
```

Comment:
This comment seems obsolete.

Food for thought: we're caching HTTP responses generated by Mimir components. Afaik Mimir doesn't set any HTTP headers with multiple values. So we could simplify our code and just use `map<string, string> headers` in protobuf definition. Although not strictly correct for general case, good-enough for Mimir use.

(I already know the answer, but worth pointing out)

Reply:
> This comment seems obsolete.
Why obsolete? 

>So we could simplify our code and just use map<string, string> headers in protobuf definition.
>(I already know the answer, but worth pointing out)
Given it's doesn't extra complexity, I prefer to keep it consistent with the `http` package where you can have multiple values, given this model is a generic one.

Reply:
> Why obsolete?

I don't see any preallocation of slice for header values. 🤔 I'm not sure what that comment refers to.

Reply:
The comment is about this pre-allocation:
```
headers := make([]*CachedHTTPHeader, 0, len(res.Header))
```

This pre-allocation is correct if we'll have 1 value for each header (common case).

Reply:
Ah, now I see what you mean. Assuming one value per header and N headers, we preallocate with N.